### PR TITLE
Use `PatternMatchAs` for mapping pattern's `rest` node

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

In the mapping pattern match, you can have a rest argument to capture the remaining fields, like:

```python
match foo:
    case {"a": 1, "b": 2, **rest}:
        pass
```

The `rest` argument has to be an identifier, and has to come last, and can't be `_`. Right now, we represent it as an identifier; this PR instead changes it to a `Pattern`, using `PatternMatchAs` with an empty `pattern` (which is equivalent to an identifier, and appears elsewhere in the pattern matching AST -- this is the pattern you get when you do `case x:` to match a wildcard and assign it to `x`).

The nice thing about this change is that the `rest` in `**rest` is now a node, not just an identifier, which I believe will significantly simplify some of the comment handling in https://github.com/astral-sh/ruff/pull/6836 (and will also enable that comment handling to look a _lot_ more like the dictionary comment handling).

## Test Plan

`cargo test`

### CODE REVIEW

File: crates/ruff_python_ast/src/nodes.rs [hunk 1]
```diff
@@ -1886,7 +1886,7 @@ pub struct PatternMatchMapping {
     pub range: TextRange,
     pub keys: Vec<Expr>,
     pub patterns: Vec<Pattern>,
-    pub rest: Option<Identifier>,
+    pub rest: Option<Box<Pattern>>,
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/nodes.rs:1889]: Why `Box<Pattern>` and not limiting it to the specific node?

**Review by MichaReiser — COMMENTED** [review]
This is interesting for two reasons:

* I don't understand why there's a dedicated node for `*name` but not `**name`
* It would be similar to `arguments` and `parameters`

The main downside that I see is that the AST now allows representing more invalid syntaxes (assuming someone mutates the AST). 

Other than that, there's a limitation that `**_` isn't valid syntax. The PEP only mentionts in the quick intro... 

> Mapping patterns: {"bandwidth": b, "latency": l} captures the "bandwidth" and "latency" values from a dict. Unlike sequence patterns, extra keys are ignored. A wildcard **rest is also supported. (But **_ would be redundant, so it is not allowed.) [[source](https://peps.python.org/pep-0636/#appendix-a-quick-intro)]

```
...     case {"a": 1, "b": 2, **_}: ...
  File "<stdin>", line 2
    case {"a": 1, "b": 2, **_}: ...
                            ^
SyntaxError: invalid syntax
```

Our parser silently parsed this before (and continues to do so now).  So this isn't a specific to this PR

Nevertheless, I'm a bit hesitant from making this change because I wonder why Python decided on their AST structure. Did you do some research to find why Python structured their AST the way they did?

**Review by konstin — APPROVED** [review]
that's a good idea

**Comment by charliermarsh** [comment]
I can't find much on it, after searching through the Discuss and the CPython repo itself, but it's hard to Google for it since there were a bunch of PEPs that went through a lot of iteration (some rejected, some superseded, etc.).

However, I did notice that in the PEP, they use `capture_pattern` for double star: https://peps.python.org/pep-0634/#mapping-patterns.

```
mapping_pattern: '{' [items_pattern] '}'
items_pattern: ','.key_value_pattern+ ','?
key_value_pattern:
    | (literal_pattern | value_pattern) ':' pattern
    | double_star_pattern
double_star_pattern: '**' capture_pattern
```

But we resolve `capture_pattern` to a `PatternMatchAs` with no `as` attribute -- that's what we match when we see `case x` in (Python also parses this as a `MatchAs` with no `as` attribute):

```python
match foo:
    case x:
        ...
```

Anyway, I'm not really sure either way on this one. I think this will make our lives easier, but then, I'd probably want to make the same change to the `as` attribute in `PatternMatchAs` (which they also represent as `as_pattern: or_pattern 'as' capture_pattern`, and we represent as an identifier).

**Comment by charliermarsh** [comment]
> Our parser silently parsed this before (and continues to do so now). So this isn't a specific to this PR

Separately: we can fix that by validating the name -- we already do that for `AsPattern`:

```rust
AsPattern: ast::Pattern = {
    <location:@L> <pattern:OrPattern> "as" <name:Identifier> <end_location:@R> =>? {
        if name.as_str() == "_" {
            Err(LexicalError {
                error: LexicalErrorType::OtherError("cannot use '_' as a target".to_string()),
                location,
            })?
        } else {
            Ok(ast::Pattern::MatchAs(
                ast::PatternMatchAs {
                   pattern: Some(Box::new(pattern)),
                   name: Some(name),
                   range: (location..end_location).into()
                },
            ))
        }
    },
}
```

**Comment by charliermarsh** [comment]
Thinking on this more, though it is a formatter-centric change, I still find it odd that `rest` is not an AST node... I think I would expect it to be.

**Comment by MichaReiser** [comment]
> Thinking on this more, though it is a formatter-centric change, I still find it odd that `rest` is not an AST node... I think I would expect it to be.

Maybe Identifier should be a node? It already has a range

**Comment by charliermarsh** [comment]
That's an interesting idea, I wouldn't mind it.

**Comment by MichaReiser** [comment]
> That's an interesting idea, I wouldn't mind it.

For me, changing Identifier to a node is easier to assess than whether we should change the pattern structure. I'm just not familiar enough with its representation and I would need to spend some time to get up to speed. 

Having a dedicated node for Identifiers would move toward my long-term goal of having a `ReferenceIdentifier` (when reading a value), and a `BindingIdentifier` (when declaring or re-declaring a binding). Having separate nodes should can be neat when querying only uses or writes. 

In the meantime, having `Identifier` as a node could be neat for other use cases too: E.g. finding all used names in a program. 

I'll leave it up to you to decide whether you want to change pattern or identifier because you have more context than I.

**Comment by charliermarsh** [comment]
@MichaReiser - That makes sense, I will convert to draft until I have time to try it.

**Comment by charliermarsh** [comment]
Not a priority right now, will revisit eventually.

**Pull request closed without merging**